src/slist.c

```diff
@@ -1 +1,297 @@
+		while (current_item != NULL)
+		{
+			bool continue_processing = false;
+
+			/* Codes_SRS_SLIST_07_029: [ `slist_remove_if` shall determine whether an entry satisfies the condition criteria by invoking the condition function for that entry. ]*/
+			if (condition_function(current_item, match_context, &continue_processing) == true)
+			{
+				/* Codes_SRS_SLIST_07_030: [ If the `condition_function` returns `true`, `slist_remove_if` shall consider that entry as to be removed. ]*/
+				if (prev_item != NULL)
+				{
+					prev_item->next = current_item->next;
+				}
+				else
+				{
+					curr_list_head = curr_list_head->next;
+				}
+			}
+			else
+			{
+				/* Codes_SRS_SLIST_07_031: [ If the `condition_function` returns `false`, `slist_remove_if` shall consider that entry as not to be removed. ]*/
+				prev_item = current_item;
+			}
+
+			/* Codes_SRS_SLIST_07_032: [ If the `condition_function` returns `continue_processing` as `false`, `slist_remove_if` shall stop iterating through the list and return. ]*/
+			if (continue_processing == false)
+			{
+				break;
+			}
+
+			current_item = current_item->next;
+		}
+		result = curr_list_head;
+	}
+
+	return result;
+}
+
+int slist_for_each(PSINGLYLINKEDLIST_ENTRY list_head, SLIST_ACTION_FUNCTION action_function, const void* action_context)
+{
+	int result;
+
+	if ((list_head == NULL) ||
+		(action_function == NULL))
+	{
+		/* Codes_SRS_SLIST_07_033: [ If `list_head` is `NULL`, `slist_for_each` shall fail and return a non - zero value. ]*/
+		/* Codes_SRS_SLIST_07_034: [ If `action_function` is `NULL`, `slist_for_each` shall fail and return a non - zero value. ]*/
+		LogError("Invalid argument (list_head=%p, action_function=%p)", list_head, action_function);
+		result = MU_FAILURE;
+	}
+	else
+	{
+		PSINGLYLINKEDLIST_ENTRY current_item = list_head;
+
+		/* Codes_SRS_SLIST_07_035: [ `slist_for_each` shall iterate through all entries in the list, invoke `action_function` for each one of themand return zero on success. ]*/
+		while (current_item != NULL)
+		{
+			bool continue_processing = false;
+
+			if (action_function(current_item, action_context, &continue_processing) != 0)
+			{
+				/* Codes_SRS_SLIST_07_036: [ If the `action_function` fails, `slist_for_each` shall fail and return a non - zero value. ]*/
+				LogError("failure in actionFunction(current_item = %p, action_context = %p, &continue_processing = %p)", current_item, action_context, &continue_processing);
+				result = MU_FAILURE;
+				break;
+			}
+			else
+			{
+				if (continue_processing == false)
+				{
+					/* Codes_SRS_SLIST_07_037: [ If the `action_function` returns `continue_processing` as `false`, `slist_for_each` shall stop iterating through the list and return. ]*/
+					result = 0;
+					break;
+				}
+			}
+			current_item = current_item->next;
+		}
+		result = 0;
+	}
+
+	return result;
+}
```

Comment:
Just a thought: how user will create the PSINGLYLINKEDLIST_ENTRY.How about adding a function (slist_create) that would create a list_head with next pointer set to NULL?